RUFAS/report_generator.py

```diff
@@ -0,0 +1,292 @@
+
+
+def sd_aggregator(data: List[float]) -> float:
+    """
+    Calculates the standard deviation of a list of numbers.
+
+    Parameters
+    ----------
+    data : List[float]
+        A list of numbers whose standard deviation is to be calculated.
+
+    Returns
+    -------
+    float
+        The standard deviation of the input numbers.
+    """
+    mean = average_aggregator(data)
+    return (sum((x - mean) ** 2 for x in data) / len(data)) ** 0.5 if data else 0
+
+
+def sum_aggregator(data: List[float]) -> float:
+    """
+    Returns the sum of a list of numbers.
+
+    Parameters
+    ----------
+    data : List[float]
+        A list of numbers whose sum is to be calculated.
+
+    Returns
+    -------
+    float
+        The sum of the input numbers.
+    """
+    return sum(data)
+
+
+def subtraction_aggregator(data: List[float]) -> float:
+    """
+    Subtracts each subsequent number in the list from the first number.
+
+    Parameters
+    ----------
+    data : List[float]
+        A list of numbers for the subtraction operation.
+
+    Returns
+    -------
+    float
+        The result of subtracting each subsequent number from the first number.
+        Returns None if the list is empty or has only one element.
+    """
+    if len(data) < 2:
+        return None
+    result = data[0]
+    for num in data[1:]:
+        result -= num
+    return result
+
+
+AGGREGATION_FUNCTIONS: Dict[str, Callable[[List[float]], float]] = {
+    "average": average_aggregator,
+    "division": division_aggregator,
+    "product": product_aggregator,
+    "SD": sd_aggregator,
+    "sum": sum_aggregator,
+    "subtraction": subtraction_aggregator
+}
+
+
+class ReportGenerator:
+    def generate_report(
+        self,
+        filtered_pool: Dict[str, Dict[str, List[Any]]],
+        filter_content: Dict[str, str | int],
+    ) -> List[float]:
+        """
+        Generates a report based on filtered data and aggregation criteria.
+
+        This method processes filtered data and performs specified horizontal and/or vertical aggregations
+        based on the instructions provided in filter_content.
+
+        Parameters
+        ----------
+        filtered_pool : Dict[str, Dict[str, List[Any]]]
+            The data pool from which the report is to be generated, structured as a dictionary.
+            Has the same structure of OutputManager's variables pool.
+
+        filter_content : Dict[str, str | int]
+            A dictionary containing filter criteria and aggregation instructions.
+
+        Returns
+        -------
+        List[float]
+            The aggregated report data as a list.
+
+        Raises
+        ------
+        ValueError
+            If the report data is empty or if the necessary aggregation keys are not found in filter_content.
+        """
+        selected_variables = filter_content.get("variables")
+        slice_start = filter_content.get("slice_start", 0)
+        slice_end = filter_content.get("slice_end", 0)
```

Comment:
What about defaulting to -1 instead, and possibly also removing the rule in `_prepare_report_data()` that 0 indicates the position of the final element? Making 0 map to the final position of a list definitely makes the code more resilient, but also makes it more confusing.
```suggestion
        slice_end = filter_content.get("slice_end", -1)
```

Reply:
I agree that this is not very clean and I am open to improvements. But `-1` won't work. The `end` in python is always exclusive, meaning this code:
```
x=[1, 2, 3, 4, 5]
print(x[:-1])
```
prints:
```
[1, 2, 3, 4]
````

Reply:
Good catch, I should have tried it before I suggested it. What about using `None` instead of -1? (That one I did try, seems like it would do the trick)

Reply:
Yes, it works! thanks for the suggestion